### PR TITLE
Fix image interpolation

### PR DESCRIPTION
**Pull request by tacaswell**
This is an attempt at  #8631 and  #6952

I expect this to fail.  Already updated a bunch of images that had minor changes.  There are a few with big changes that I still need to look into.

 - [x] re-visit rotate image and get rid of black pixels on edge
 - [x] drop attempt to fix python2
 - [x] add comments about ringing into test code
 - [x] ~~look into scaling to a bit _less_ than full range to prevent speckles near vmin/vmax~~
 - [x] test flat cases near vmin or vmax\
 - [x] regenerate all test images with current defaults

### CODE REVIEW

**Comment by anntzer** [comment]
In rotate_image.png, the edge of the image seems to follow the middle of the dashes more closely in the original version.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -356,58 +356,87 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
+                _image.resample(A_scaled, A_resampled,
+                                t,
+                                _interpd_[self.get_interpolation()],
+                                self.get_resample(), 1.0,
+                                self.get_filternorm() or 0.0,
+                                self.get_filterrad() or 0.0)
+
+                # we are done with A_scaled now, remove from namespace to be sure!
+                del A_scaled
+                # un-scale the resampled data to approximatly the
```
Comment (dopplershift) [inline lib/matplotlib/image.py]: approximatly -> approximate

File: lib/matplotlib/image.py [hunk 2]
```diff
@@ -356,58 +356,87 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
+                out_mask[:] = np.nan
+                _image.resample(mask, out_mask,
+                                t,
+                                _interpd_[self.get_interpolation()],
+                                True, 1,
+                                self.get_filternorm() or 0.0,
+                                self.get_filterrad() or 0.0)
+                # we are done with the mask, delete from namespace to be sure!
+                del mask
+                # Agg tells us a pixel has no value not setting a value into it
```
Comment (dopplershift) [inline lib/matplotlib/image.py]: I think there's a word missing in this comment.

**Comment by tacaswell** [comment]
I anticipate re-writing history here before it gets merged.

**Comment by tacaswell** [comment]
I see some failures locally that I do not understand, let see if we see them on travis.

Still have to sort out those edge effects.

History will get re-written at least once more before this is ready to merge.

**Comment by tacaswell** [comment]
I am actually comfertable that the edges are the correct behavior.  If you look at the zoomed in bottom 3x3 area

so

you can see that the out of range area is against the edge, the bottom right is the highest value which is equal to vmax of the norm.

It is correct, from the interpolations point of view, that the pixel past the last point (which it view as being at the center of that pixel) are of greater value (which are then correctly marked as out of range).

**Comment by tacaswell** [comment]
Todos:

 - [ ] re-visit rotate image and get rid of black pixels on edge
 - [ ] drop attempt to fix python2
 - [ ] add comments about ringing into test code
 - [ ] look into scaling to a bit _less_ than full range to prevent speckles near vmin/vmax
 - [ ] test flat cases near vmin or vmax

**Comment by tacaswell** [comment]
```python
from copy import copy
import matplotlib.image as mimage
def interpolation_grid(data, vmin=None, vmax=None):
    cm = copy(plt.get_cmap('viridis'))
    cm.set_over('r')
    cm.set_under('b')
    cm.set_bad('k')
    
    fig, ax_grid = plt.subplots(3, 6, sharex=True, sharey=True)

    ret = []
    for interp, ax in zip(sorted(mimage._interpd_), ax_grid.ravel()):
        ax.set_title(interp)
        im = ax.imshow(data, cmap=cm, interpolation=interp)
        ax.axis('off')
        im.set_clim(vmin, vmax)
        ret.append(im)
    return ret


```

```python
ims_a = interpolation_grid([[0, 1], [1, 0]], 0, 1)
ims_b = interpolation_grid(np.tile([[0, 1], [1, 0]], (10, 10)), 0, 1)
```

I am satisfied that the speckles are edge effects on the higher-order interpolators 

```
[im.set_clim(0, 1.00000005) for im in ims_b]
```

That is the smallest upper limit that shows the speckles on my system which gives us an idea of the level of the numeric instability.  I don't think that it is worth playing games to stay just under it as I suspect it will vary between systems.

If we do anything about this we should bump the Autoscales out a bit.

**Pull request merged**